Add translate SparkSQL function

PHILO-HE opened the pull request.

PHILO-HE commented:
@mbasmanova, I am proposing this patch for spark translate function. Could you help review it? Thanks!

mbasmanova commented:
@PHILO-HE Hi, which company are you from and which project are you working on? Please, review contributing guidelines and update the PR as necessary. For example, I see that documentation for the new function is missing.

PHILO-HE commented:
> @PHILO-HE Hi, which company are you from and which project are you working on? Please, review contributing guidelines and update the PR as necessary. For example, I see that documentation for the new function is missing.

Hi @mbasmanova, I'm from Intel Gluten team. Just updated doc for this new function. Thanks!

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -150,6 +150,12 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT substring('Spark SQL', -20, 3); -- ""
 
+.. spark:function:: translate(string, matchStr, replaceStr) -> varchar
+    Returns a new translated string. It translates the character in ``string`` by a
```
Would you generate the docs and verify that they look nice. Don't you need an empty line here?

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -150,6 +150,12 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT substring('Spark SQL', -10, 3); -- "Sp"
         SELECT substring('Spark SQL', -20, 3); -- ""
 
+.. spark:function:: translate(string, matchStr, replaceStr) -> varchar
```
matchStr -> match 
replaceStr -> replace

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -150,6 +150,12 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT substring('Spark SQL', -20, 3); -- ""
 
+.. spark:function:: translate(string, matchStr, replaceStr) -> varchar
+    Returns a new translated string. It translates the character in ``string`` by a
```
Would you share a link to Spark's documentation for this function?

PHILO-HE replied:
A simple doc [here](https://spark.apache.org/docs/latest/api/sql/index.html#translate). Source code comment [here](
https://github.com/apache/spark/blob/1a5b9f295a623901100b6f41e785138fba0bc156/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L863).
Assume no need to paste the links in velox doc.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  using MapType = std::unordered_map<StringView, StringView>;
```
std::unordered_map is slow; use F14 instead

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  MapType dict_;
+  bool isConstantTranslateDict = false;
+
+  FOLLY_ALWAYS_INLINE MapType buildDict(
```
do not abbreviate: buildDictionary, match, replace. Please, check out coding style guide and update.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+    while (i < matchStr.size()) {
+      StringView replacePiece;
+      if (j >= replaceStr.size()) {
+        replacePiece = StringView("");
```
This logic suggests that 'replace' string may be shorter than 'match' string, in which case remaining characters in 'match' are removed from the input string. Is this how Spark's function behaves? If so, please, document this behavior. It would be nice to add some examples to documentation to illustrate common cases and special cases.

PHILO-HE replied:
Yes, this is exactly spark's behavior. Just documented. Thank!

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+      const arg_type<Varchar>& matchStr,
+      const arg_type<Varchar>& replaceStr) {
+    MapType dict;
+    int i = 0, j = 0;
```
Put each variable on its own line. Check coding style.

PHILO-HE replied:
Fixed.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+      }
+      int matchCharCount = utf8proc_char_length(matchStr.data() + i);
+      StringView matchPiece = StringView(matchStr.data() + i, matchCharCount);
+      dict.try_emplace(matchPiece, replacePiece);
```
Should we check if there are duplicates and throw? What does Spark do in this case? Let's document.

PHILO-HE replied:
Spark only considers the firstly occurred character in match string. So we dismiss the match character and its corresponding replace character if the match character already exists in the built dictionary.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+      if (j >= replaceStr.size()) {
+        replacePiece = StringView("");
+      } else {
+        int replaceCharCount = utf8proc_char_length(replaceStr.data() + j);
```
This is slow when input is ASCII. Let's add fast path for ASCII inputs.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+    }
+    // No need to do the replacement.
+    if (dict_.empty()) {
+      result.setNoCopy(StringView(input.data(), input.size()));
```
Why setNoCopy? I imagine we need to copy the string, no?

PHILO-HE replied:
For this case, just re-use the input string as result. So I use setNoCopy. Am I missing something? Thanks!

mbasmanova replied:
To reuse input strings, one needs to set reuse_strings_from_arg to tell the framework that input string buffers need to be added to the result vector. Otherwise, you'll end up having a result vector referring to memory that it doesn't own.

https://facebookincubator.github.io/velox/develop/scalar-functions.html#zero-copy-string-result

PHILO-HE replied:
Thanks for letting me know this. Yes, I missed this key point. Just added the reuse flag.
A question. For such function that produces no-copy result for some cases and doesn't for other cases, is it a best practice to still set the reuse flag? As with it, the framework will not let the input buffer go away prematurely, even for copy cases.

mbasmanova replied:
In this case, I think it would be rare that input strings can be reused. Hence, I'd suggest not to set reuse flag and not to use setNoCopy. Otherwise. you'll end up keeping unused strings buffers in memory for longer than needed and waste memory.

PHILO-HE replied:
Absolutely agree. That reuse case is very rare in real workload. I just removed the usage of `setNoCopy`, as well as the reuse flag. Thanks!

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+    }
+    int k = 0;
+    while (k < input.size()) {
+      int inputCharCount = utf8proc_char_length(input.data() + k);
```
Let's add fast ASCII path. It would be nice to add a benchmark to compare fast ASCII path with regular path.

PHILO-HE replied:
Yes, it deserves to add a fast path for ASCII. But in `initialize()` method, we still uses the regular path to build dictionary, assuming there is no similar function interface for implementation as `callAscii`.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+    while (k < input.size()) {
+      int inputCharCount = utf8proc_char_length(input.data() + k);
+      StringView inputPiece = StringView(input.data() + k, inputCharCount);
+      if (dict_.find(inputPiece) == dict_.end()) {
```
This way you are going to lookup twice. First here, then in "dict_[inputPiece". Use find method instead.

PHILO-HE replied:
Yes, just corrected.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+      return;
+    }
+    int k = 0;
+    while (k < input.size()) {
```
It looks like it would be helpful to add a function that loops over characters of a string and applies a lambda to each character. Or, create an iterator that returns next character.

PHILO-HE replied:
Yes, we can move this part into a function with utf-8/ ascii considered. Please take a review to see whether I fully caught your point. Thanks!

mbasmanova reviewed (commented):
@PHILO-HE Some comments.

PHILO-HE commented on velox/docs/functions/spark/coverage.rst:
```diff
@@ -104,7 +104,7 @@
     :spark:func:`asinh`                        degrees                                    ln                                         reflect                                    :spark:func:`transform`                        last_value                                                                              
     assert_true                                dense_rank                                 locate                                     regexp                                     transform_keys                                 max                                                                                     
     atan                                       div                                        log                                        :spark:func:`regexp_extract`               transform_values                               max_by                                                                                  
-    atan2                                      double                                     log10                                      regexp_extract_all                         translate                                      mean                                                                                    
```
Makes sense. I just reverted the change. Thanks!

PHILO-HE commented:
@mbasmanova, could you please review this patch again? Thanks! A CI failure looks not relevant to this PR.

mbasmanova commented:
CC @kgpai @assignUser Folks, is the CI failure here expected?

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,122 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  FOLLY_ALWAYS_INLINE void doTranslate(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      std::function<int(const char* input)> charCounter) {
```
How fast is this implementation for all-ASCII input? I think it may be not very fast. I suggest to create a separate implementation for all-ASCII input and experiment with using an array of chars for dictionary.

PHILO-HE replied:
Hi @mbasmanova, currently, we use a unified map (with StringView or the suggested std::string), which can be created at `initialize()` for constant args, regardless of ASCII input or not. It looks there is no way to recognize ASCII input at `initialize()` and then do the suggested optimizations. So I'm assuming your suggested optimizations are applicable to non-constant path. But, according to the feedback from our customers, it is more common that constant args (`replace`/`match`) are used for this functions. Do we still need to do the optimization for non-constant path? Please correct me if I thought wrongly. Thanks!

mbasmanova replied:
I agree that constant dictionary with all-ASCII inputs is the most common path. This is the one we should optimize for. You can easily check if all values in dictionary are all-ASCII, no?

PHILO-HE replied:
In the latest commit, I separated the implementations for all-Ascii path and the others. And a map with char type used is for all-Ascii input. In addition, building the map is moving from `initialize` to `call`/`callAscii` for constant args and a flag is introduced to mark whether it is already built for constant case. Thus, we can build the corresponding kind of map, regardless of constant args or not.
With your suggested optimization, around 50% time can be saved for all-Ascii input in a simple test (input string with 10k ascii characters, 100k iterations), compared with the previous implementation.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,122 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+    return [](const char*) { return 1; };
+  }
+
+  using MapType = folly::F14FastMap<StringView, StringView>;
```
StringView is just a pointer. If the memory referenced by it is deallocated, this map will get corrupted.

For all-ASCII input, this can be just a map from char to char (or a plan array indexed by ASCII code of 'char')). For non-ASCII input you may need to store std::strings.

PHILO-HE replied:
> StringView is just a pointer. If the memory referenced by it is deallocated, this map will get corrupted.

Hi @mbasmanova, I think the map's key/value can only be accessed before the referred memory (for `match`/`replace`) is deallocated. Is there any possible case I am missing? Thanks!

mbasmanova replied:
It is not guaranteed that StringViews passed to initialize() stay alive for as along as the function is being used. These can go away.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,122 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+      StringView inputPiece = StringView(input.data() + k, inputCharCount);
+      auto it = dictionary_.find(inputPiece);
+      if (it == dictionary_.end()) {
+        result.append(inputPiece);
```
Wouldn't appending one character at a time be very slow?

PHILO-HE replied:
Just aware of this issue. Fixed in the latest commit by reserving capacity.

mbasmanova reviewed (commented):
Some comments.

PHILO-HE commented:
@mbasmanova, please review again. Thanks!

kgpai commented:
@mbasmanova I suppose any CI failure was transient , if not please post a link so we can investigate more.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,167 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  static constexpr bool is_default_ascii_behavior = true;
+
+  folly::F14FastMap<std::string, std::string> dictionaryUtf8_;
+  folly::F14FastMap<char, char> dictionaryAscii_;
```
nit: consider renaming to asciiDictionary_ and unicodeDictionary_. This way the reader gets to the important information faster (ascii vs. unicode) .

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,167 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  folly::F14FastMap<std::string, std::string> dictionaryUtf8_;
+  folly::F14FastMap<char, char> dictionaryAscii_;
+
+  bool isConstantTranslateDictionary = false;
```
isConstantDictionary_; make sure to add trailing underscore

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,167 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  folly::F14FastMap<char, char> dictionaryAscii_;
+
+  bool isConstantTranslateDictionary = false;
+  bool isConstantTranslateDictionaryBuilt = false;
```
This is not needed. Use isAscii function from ./velox/functions/lib/string/StringCore.h

```
/// Check if a given string is ascii
static bool isAscii(const char* str, size_t length);
```

PHILO-HE replied:
Hi @mbasmanova, `isConstantTranslateDictionaryBuilt` is used for marking whether the dictionary is built for constant arg case. Since it it impossible to identify all-ascii input in `initialize()`, we uniformly build in the corresponding kind of dictionary call/callAscii. Am I missing something? Thanks!

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,167 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  bool isConstantTranslateDictionaryBuilt = false;
+
+  FOLLY_ALWAYS_INLINE folly::F14FastMap<std::string, std::string>
+  buildDictionaryUtf8(
```
nit: buildUnicodeDictionary

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,167 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+      auto it = dictionaryAscii_.find(inputChar);
+      if (it == dictionaryAscii_.end()) {
+        std::memcpy(result.data() + i, &inputChar, 1);
+        i = i + 1;
```
++i

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -583,4 +583,167 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+    result.reserve(input.size());
+    int i = 0;
+    int k = 0;
+    while (k < input.size()) {
```
use a for loop

mbasmanova reviewed (commented):
Some further comments. I made each comment only once, but many of these applies to multiple places in the code. Please, make sure to fix all instances.

PHILO-HE commented:
@mbasmanova, could you please further review this PR? Thanks so much!

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+  folly::F14FastMap<std::string, std::string> unicodeDictionary_;
+  folly::F14FastMap<char, char> asciiDictionary_;
+
+  bool isConstantTranslateDictionary_ = false;
```
nit: isConstantTranslateDictionary_ -> isConstantDictionary_

ditto isConstantTranslateDictionaryBuilt_

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+  bool isConstantTranslateDictionary_ = false;
+  bool isConstantTranslateDictionaryBuilt_ = false;
+
+  FOLLY_ALWAYS_INLINE folly::F14FastMap<std::string, std::string>
```
FOLLY_ALWAYS_INLINE is not needed; no point inlining such a large function

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+    int i = 0;
+    int j = 0;
+    while (i < match.size()) {
+      std::string replacePiece;
```
why 'Piece'? maybe, replaceChar

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+    return dictionary;
+  }
+
+  FOLLY_ALWAYS_INLINE folly::F14FastMap<char, char> buildAsciiDictionary(
```
> FOLLY_ALWAYS_INLINE

not needed

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    if (!isConstantTranslateDictionary_ ||
+        !isConstantTranslateDictionaryBuilt_) {
```
This logic doesn't seem right. It is possible that first batch of data is all-ASCII, but next is not. In this case, on first batch we'll build asciiDictionary and set isConstantTranslateDictionaryBuilt_ to true. On next batch, we'll assume unicodeDictionary_ is built, but it will be empty.

Would be nice to add such a test case as well as reverse (unicode batch followed by ascii batch).

PHILO-HE replied:
@mbasmanova, thanks for your insightful comment. Yes, it indeed has a critical issue in firstly tackling ascii batch, then unicode batch, or reversely. Just fixed by using two independent flags to track whether ascii/unicode dictionary is built. And a simple test was also just added to cover such case. Thanks so much!

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,155 @@ struct LeftFunction {
+    int k = 0;
+    while (k < input.size()) {
+      int inputCharLength = utf8proc_char_length(input.data() + k);
+      auto inputPiece = std::string(input.data() + k, inputCharLength);
```
inputPiece -> inputChar

mbasmanova reviewed (commented):
Some further comments.

PHILO-HE commented on velox/functions/sparksql/String.h:
```diff
@@ -638,4 +638,153 @@ struct LeftFunction {
+      const arg_type<Varchar>& replace) {
+    if (!isConstantDictionary_ || !isConstantUnicodeDictionaryBuilt_) {
+      unicodeDictionary_ = buildUnicodeDictionary(match, replace);
+      if (!isConstantUnicodeDictionaryBuilt_) {
```
Hi @mbasmanova, if dictionary is constant, it will be built only if it is not built yet, i.e., `isConstantUnicodeDictionaryBuilt_=false` here. Not building it for each row.

mbasmanova replied:
Sorry, I meant to ask what happens if dictionary is not constant.

PHILO-HE replied:
If dictionary is not constant, which means `match` & `replace` are not constant, we have to build it for each row. Then, `isConstantUnicodeDictionaryBuilt_` means nothing.

mbasmanova replied:
Do we have a test case that verifies that non-constant dictionary use case works properly?

PHILO-HE replied:
Hi @mbasmanova, I just added some tests for covering non-constant cases. Thanks!

mbasmanova replied:
I got confused by the code setting isConstantUnicodeDictionaryBuilt_ to true even when dictionary is not constant. It happens to work, but I expect future readers to be confused as well. Consider changing the logic to 

```
if (isConstantDictionary_) {
   isConstantUnicodeDictionaryBuilt_ = true;
}
```

PHILO-HE replied:
Agree. It can be logically more clear. Just fixed. Thanks!

mbasmanova commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -497,5 +522,105 @@ TEST_F(StringTest, left) {
+  EXPECT_EQ(translate("abc", "\u2028", std::nullopt), std::nullopt);
+  EXPECT_EQ(translate(std::nullopt, "\u2028", "\u2029"), std::nullopt);
+
+  // Test with flat vector.
```
This comment as well as method name testTranslateFlatVector suggests that there is something special about flat vectors that requires specific testing. I believe this is not the case. In fact, the function implementation is agnostic to vector encoding as it receives values one row at at time and doesn't have any handing of encodings. The fragment handles the encoding.

We do need to test fast ASCII paths as well as cases where we have a mix of all-ASCII and non-all-ASCII batches of data.

Would it be possible to refactor this tests to make it clear what is being tested?

PHILO-HE replied:
I just refactored the test code. And the encoding setting for test batch is removed. It may look more clear to reader now. 

BTW, in the latest implementation for translate, I introduced `std::optional` for the two dictionaries and removed the two flags that used for marking whether they are built.

PHILO-HE commented:
> @PHILO-HE There are CI failures. Would you rebase to see if they clear? CC: @xiaoxmeng
> 
> Also, please, squash comments. There are 35 commits now. GitHub gets slow when there are large number of commits.

@mbasmanova, I have squashed the commits and rebased the code. All checks have passed now. If you have any further comment, please let me know. Thanks!

mbasmanova commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -497,5 +520,91 @@ TEST_F(StringTest, left) {
+  EXPECT_EQ(translate("abc", "\u2028", std::nullopt), std::nullopt);
+  EXPECT_EQ(translate(std::nullopt, "\u2028", "\u2029"), std::nullopt);
+
+  // Test constant dictionary cases.
```
Thank you for iterating on the tests.

It is easier to debug test failures when tests are small. Hence, I suggest to break up this test into 3. I then suggest to remove helper methods and use lambdas. This will make tests shorter and all test code will be in one places making it easier to read. Here is an example:

```
TEST_F(StringTest, translateConstantMatch) {
  auto rowType = ROW({{"c0", VARCHAR()}});
  auto exprSet = compileExpression("translate(c0, 'ab', '12')", rowType);

  auto testTranslate = [&](const auto& input, const auto& expected) {
    auto result = evaluate(*exprSet, makeRowVector({input}));
    velox::test::assertEqualVectors(expected, result);
  };

  auto input = makeFlatVector<std::string>({"abcd", "cdab"});
  auto expected = makeFlatVector<std::string>({"12cd", "cd12"});
  testTranslate(input, expected);

  // Uses unicode batch as the next input.
  input = makeFlatVector<std::string>({"abåæçè", "åæçèab"});
  expected = makeFlatVector<std::string>({"12åæçè", "åæçè12"});
  testTranslate(input, expected);
}
```

PHILO-HE replied:
Yes, it makes sense. Just did some further code refactor according to your nice example. Thanks!

mbasmanova commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -163,6 +163,14 @@ class StringTest : public SparkFunctionBaseTest {
         len);
   }
+
+  std::optional<std::string> translate(
```
Let's move it into 'translate' test as a lambda as well.

PHILO-HE replied:
Done. Thanks!

mbasmanova commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -497,5 +505,73 @@ TEST_F(StringTest, left) {
+  };
+
+  // All inputs are ascii encoded.
+  auto input1 = makeFlatVector<std::string>({"abcd", "cdab"});
```
variable names a1, a2, a3 are an anti-pattern; let's rename to input, match and replace

mbasmanova reviewed (commented):
@PHILO-HE Thank you for revising. Looks great % a couple of nits.

mbasmanova reviewed (approved):
Thank you. Looks great.
